### DISCUSSION TITLE
using opengrok/docker image for perforce repository

### DISCUSSION THREAD

**Discussion by kee98** [oracle/opengrok#4090] [Q&A]
I am working on setting up new opengrok instance for our organization and as obvious my preference is using the Docker version. 
I followed all the steps and able to setup the UI. Even I am able to see the perforce project under a "Project" section on the home dashboard. Search, explore repo etc works well. However, History and Annotation links are always disabled. Is this because, docker image doesn't support perforce? If so, any suggestion if we can tweak some code (patch to make it work) to make it work for perforce. I also see perforce folder not created under /opengrok/data/historycache folder (whereas for git projects, it has created the folder).

Thanks in advance.

**Comment by vladak** [comment]
Yep, Perforce is not present in the image, see https://github.com/oracle/opengrok/blob/master/Dockerfile and yes, the history/annotations are greyed out if the repository cannot provide them in any way.

Now, I am not sure whether you can actually distribute Perforce in a Docker image listed on Docker Hub.

**Reply by kee98** [reply]
Thanks for response. That means I need to fall back to use traditional VM way to deploy. However, any plan to support perforce in future for docker image? Because, I don't think it should be that complicated as opengrok itself supports perforce.

**Reply by vladak** [reply]
Not at the moment. As you note, this should be a matter of adding Perforce installation to the Docker image. PRs are welcome.

**Reply by kee98** [reply]
Sure, I will open a PR for that. Before that I was trying to make this work for me if its not complex to achieve. I added below code to Dockerfile for perforce installation but History still not generated. I will try to debug this more and update this thread meanwhile if you have any more info about what else I might need to change then that will be helpful.

```
RUN apt-get update && \
    apt-get install -y wget gnupg2

RUN wget -qO - https://package.perforce.com/perforce.pubkey | apt-key add -

RUN echo 'deb http://package.perforce.com/apt/ubuntu bionic release' > /etc/apt/sources.list.d/perforce.list

RUN apt-get update && \
    apt-get install -y helix-p4d
```

**Reply by kee98** [reply]
I am getting error `WARNING: History cache is enabled however the list of repositories is empty.` for perforce projects. However, for git projects, its able to created history cache.

**Reply by kee98** [reply]
I see this command its running to index (and failing to create history with above warning):

`opengrok-reindex-project --printoutput --jar /opengrok/lib/opengrok.jar -U http://localhost:8080/ -P perforce-xxx.stage.exa -- -r dirbased -G -m 256 --leadingWildCards on -c /usr/local/bin/ctags -U http://localhost:8080/ -H perforce-xxx.stage.exa`

I tried passing -S and running it manually but of no use.

**Reply by kee98** [reply]
Tried running it from indexer directly but it seems somehow opengrok is not recognizing perforce projects:

`java -Dorg.opengrok.indexer.history.Perforce=/opengrok/dist/lib/ -jar /opengrok/dist/lib/opengrok.jar -H -S -P -s /opengrok/src -d /opengrok/data ...`

Log snippet:

```
INFO: Scanning for repositories in [/opengrok/src/perforce-xxx.stage.abc/abc, /opengrok/src/perforce-xxx.abc/.p4config] (down to 2 levels below source root)
Nov 07, 2022 5:06:19 PM org.opengrok.indexer.util.Statistics logIt
INFO: Done scanning for repositories, found 0 repositories (took 26 ms)
```


Can you please let me know if we are missing something small?

**Reply by vladak** [reply]
> Tried running it from indexer directly but it seems somehow opengrok is not recognizing perforce projects:
> 
> `java -Dorg.opengrok.indexer.history.Perforce=/opengrok/dist/lib/ -jar /opengrok/dist/lib/opengrok.jar -H -S -P -s /opengrok/src -d /opengrok/data ...`

The value of the `org.opengrok.indexer.history.Perforce` property has to be the full path to the actual `p4` binary.

**Reply by vladak** [reply]
To make it a little easier to detect such problems, I created #4094.

**Reply by vladak** [reply]
There is also special handling done to find out if a file/directory is part of a Perforce repository. I don't have experience with Perforce, so cannot tell what needs to be done for the repositories. See the code in https://github.com/oracle/opengrok/blob/c1a9151373191b459f78afaf4dcd55e29ca09383/opengrok-indexer/src/main/java/org/opengrok/indexer/history/PerforceRepository.java#L145-L152

**Reply by kee98** [reply]
> The value of the org.opengrok.indexer.history.Perforce property has to be the full path to the actual p4 binary.

Thanks, it seems this was useful to some extent. Now I can see historycache created for perforce project, however it has only file `OpenGroklatestRev` with content commit id and following warnings in log:

`WARNING: Changes missed revision 17198 for /perforce-xxx.abc/proj/main/webapps/watchbuild/views.py
`
May be its waiting for indexer to finish before creating full cache. I will wait and will try to check this more tomorrow.

**Reply by vladak** [reply]
What is in the indexer logs actually ?

**Reply by kee98** [reply]
I just see a lot of similar message in indexer log:

`WARNING: Fields [hist] will be ignored because they were not found in index directory MMapDirectory@/opengrok/data/index/perforce-xxx.abc.stage lockFactory=org.apache.lucene.store.NativeFSLockFactory@6a6dda4b`

**Reply by kee98** [reply]
Basically after manually going inside the perforce directory and running the command:
`java -Dorg.opengrok.indexer.history.Perforce=/opengrok/dist/lib/ ...`

it creates history direcory but only with file name `OpenGroklatestRev` which contains revision id. and nothing else.

**Reply by vladak** [reply]
> I just see a lot of similar message in indexer log:
> 
> `WARNING: Fields [hist] will be ignored because they were not found in index directory MMapDirectory@/opengrok/data/index/perforce-xxx.abc.stage lockFactory=org.apache.lucene.store.NativeFSLockFactory@6a6dda4b`

This is from the webapp log, not indexer log. Also, at this point such warning is innocuous.

**Comment by vladak** [comment]
What is the return code of '/usr/bin/p4 help' ?

________________________________________
From: kee98 ***@***.***>
Sent: Tuesday, November 8, 2022 5:46 AM
To: oracle/opengrok
Cc: Vladimir Kotal; Comment
Subject: [External] : Re: [oracle/opengrok] using opengrok/docker image for perforce repository (Discussion #4090)

I tried merging your change in my local and build the image. Now I can see following error while running this command:

java -Dorg.opengrok.indexer.history.Perforce=/usr/bin/p4 -jar /opengrok/lib/opengrok.jar -H -S -P -s /opengrok/src -d /opengrok/data
log:

WARNING: Command '/usr/bin/p4' does not work. Some operations with Perforce repositories will fail as a result.

However, /usr/bin/p4 exists:

# /usr/bin/p4 info
User name: name
Client name: opengrok-dev-stage-xxx-02
Client host: opengrok-c5898674f-m2wvf
...
...

**Reply by kee98** [reply]
If I run it from /opengrok/src/<perforce_directory/ then I get 0 return code. However, from other directory, I get 1 return code.

**Reply by vladak** [reply]
Okay, that explains the failure. Will change the Perforce check from `p4 help` to `p4 -V` as the latter seems to exit with 0 no matter in which directory it is executed.

**Reply by kee98** [reply]
this is my complete log, if that helps. It seems to be picking the directory right, but not generating history instead just creates OpenGroklatestRev file.

When I run history command for first time, I get several warning like: 

PerforceHistoryParser.parseTruncatedFileLog: Changes missed revision xxxx for file <>/build.py

and then for any subsequent run it shows below logs.  That means, its picking the project fine, but somehow (some configuration?) its only generating OpenGroklatestRev file and not actualy history.

```
Exuberant Ctags 5.8, Copyright (C) 1996-2009 Darren Hiebert
  Compiled: Nov  3 2022, 14:18:55
  URL: https://ctags.io/
  Optional compiled features: +wildcards, +regex, +iconv, +option-directory, +packcc, +optscript
13:03:22 INFO: Scanning for repositories in [/opengrok/src/perforce-xxx.abc.is.here] (down to 2 levels below source root)
13:03:22 WARNING: Command '/usr/bin/p4' does not work. Some operations with Perforce repositories will fail as a result.
13:03:22 INFO: Done scanning for repositories, found 1 repositories (took 490 ms)
13:03:22 INFO: Generating history cache for all repositories ...
13:03:23 INFO: Creating history cache for 1 repositories
13:03:23 INFO: Creating history cache for /opengrok/src/perforce-xxx.abc.is.here (PerforceRepository) without renamed file handling
13:03:23 INFO: Done history cache for /opengrok/src/perforce-xxx.abc.is.here (took 249 ms)
13:03:23 INFO: Done history cache for all repositories (took 397 ms)
13:03:23 INFO: Done generating history cache
13:03:23 INFO: Starting indexing
```

I get the warning for /usr/bin/p4 (even after changing `boolean works = checkCmd(repoCommand, "-V");` in `opengrok-indexer/src/main/java/org/opengrok/indexer/history/PerforceRepository.java` file.

Surprise to see there are no good/ detailed documentations for opengrok with perforce :( . If there are any, please share (even if it is for non-docker opengrok).

**Reply by kee98** [reply]
ok, some magic happened and now I can see history and annotations...not sure what exactly I did but it was just some random different things and debugging the codes. I will try this on fresh setup and update this thread if there are anything useful for opengrok community.

**Reply by vladak** [reply]
> I get the warning for /usr/bin/p4 (even after changing `boolean works = checkCmd(repoCommand, "-V");` in `opengrok-indexer/src/main/java/org/opengrok/indexer/history/PerforceRepository.java` file.

There was logic error in the code so the warning was always displayed. Fixed that in the PR.
 
> Surprise to see there are no good/ detailed documentations for opengrok with perforce :( . If there are any, please share (even if it is for non-docker opengrok).

There may be some hints in closed OpenGrok issues. I did some spelunking there however did not find anything valuable.
